docs/user_guide/ml.md

```diff
@@ -1,8 +1,254 @@
+
+## datasets:
+In the {py:meth}`~columnflow.ml.MLModel.datasets` function, you specify which datasets are important for your machine learning model and which dataset instance should be extracted from your config.
+To use this function your datasets needs to be added to your campaign, as defined by the [Order](https://python-order.readthedocs.io/en/latest/>) Module.
+An example can be found [here](https://github.com/uhh-cms/cmsdb/blob/d83fb085d6e43fe9fc362df75bbc12816b68d413/cmsdb/campaigns/run2_2018_nano_uhh_v11/top.py).
+It is recommended to return this as `set`, to prevent double counting.
+In the following example all datasets given by the {ref} external config <init_of_ml_model> are taken, but also an additional is given.
+Note how the name of each dataset is used to get an `dataset instance` from your `config instance`, this ensures that you properly use the correct dataset.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def datasets
+:end-at: return set(dataset_inst)
+```
+
+
+## produces:
+By default, intermediate columns are not saved within the columnflow framework but are filtered out afterwards.
+If you want to prevent certain columns from being filtered out, you need to tell columnflow by writing them in the {py:meth}`~columnflow.ml.MLModel.produces`function.
+This is always done by writing an iterable containing the name of the producer as string.
+More information can be found in the official documentation about [producers](building_blocks/producers).
+
+In the following example, I want to preserve the number of muons and electrons, we also want to preserve the output of the neural network.
+To avoid confusion, we are not producing the columns in this function, we only tell columnflow to not throwing them away.
+```{literalinclude} ./ml_code.py
+:language: python
+:start-at: def produces
+:end-at: return preserved_columns
+```
+ATTENTION: Sometimes one also want to preserve the used input and target features, for example to make plots.
+Remember that you already have a copy of all the input and target features in your parquet files used for training.
+In General it is not advised to create a new separate copy of these, otherwise it can be convenient to have everything within one place.
+Thus, the choice to include these columns is your choice.
+
+## uses:
+In `uses` you define the columns that are needed by your machine learning model, and there that are not filtered out.
+In this case we want to preserve the input and target features, as well as some weights:
+```{literalinclude} ./ml_code.py
+:language: python
+:pyobject: MLModel.uses
+```
+
+## output:
+In the {py:meth}`~columnflow.ml.MLModel.output` function, you define your local target directory that your current model instance will have access to.
+
+Since machine learning in columnflow uses k-fold by default, it is a good idea to have a separate directory for each fold, and this should be reflected in the {py:meth}`~columnflow.ml.MLModel.output` path.
```

Comment:
```suggestion
Since machine learning in columnflow uses k-fold cross validation by default, it is a good idea to have a separate directory for each fold, and this should be reflected in the {py:meth}`~columnflow.ml.MLModel.output` path.
```

Also: You write this as if this is a well-established fact, but I don't think so. While most people probably should use k-fold cross validation, not a lot of people do. So this should definitely be something that is stated pretty early on on this page